commit 1b6f332c3c539a45f6ed5a85897ecf8a65fbb216
Author: Lukas Tinkl <lukas@kde.org>
Date:   Mon Dec 12 20:03:43 2011 +0100

    disconnect the battery signals when it gets removed
    
    see https://bugzilla.redhat.com/show_bug.cgi?id=753429 and
    https://bugzilla.redhat.com/show_bug.cgi?id=766399

diff --git a/plasma/generic/dataengines/powermanagement/powermanagementengine.cpp b/plasma/generic/dataengines/powermanagement/powermanagementengine.cpp
index a5344da..2bee9a7 100644
--- a/plasma/generic/dataengines/powermanagement/powermanagementengine.cpp
+++ b/plasma/generic/dataengines/powermanagement/powermanagementengine.cpp
@@ -105,8 +105,7 @@ QStringList PowermanagementEngine::sources() const
 bool PowermanagementEngine::sourceRequestEvent(const QString &name)
 {
     if (name == "Battery") {
-        const QList<Solid::Device> listBattery =
-                        Solid::Device::listFromType(Solid::DeviceInterface::Battery, QString());
+        const QList<Solid::Device> listBattery = Solid::Device::listFromType(Solid::DeviceInterface::Battery);
         m_batterySources.clear();
 
         if (listBattery.isEmpty()) {
@@ -160,7 +159,7 @@ bool PowermanagementEngine::sourceRequestEvent(const QString &name)
     } else if (name == "AC Adapter") {
         bool isPlugged = false;
 
-        const QList<Solid::Device> list_ac = Solid::Device::listFromType(Solid::DeviceInterface::AcAdapter, QString());
+        const QList<Solid::Device> list_ac = Solid::Device::listFromType(Solid::DeviceInterface::AcAdapter);
         foreach (Solid::Device device_ac, list_ac) {
             Solid::AcAdapter* acadapter = device_ac.as<Solid::AcAdapter>();
             isPlugged |= acadapter->isPlugged();
@@ -232,6 +231,11 @@ void PowermanagementEngine::updateAcPlugState(bool newState)
 void PowermanagementEngine::deviceRemoved(const QString& udi)
 {
     if (m_batterySources.contains(udi)) {
+        Solid::Device device(udi);
+        Solid::Battery* battery = device.as<Solid::Battery>();
+        if (battery)
+            battery->disconnect();
+
         const QString source = m_batterySources[udi];
         m_batterySources.remove(udi);
         removeSource(source);
diff --git a/plasma/generic/dataengines/powermanagement/powermanagementengine.h b/plasma/generic/dataengines/powermanagement/powermanagementengine.h
index 4dcc0d9..96ffd45 100644
--- a/plasma/generic/dataengines/powermanagement/powermanagementengine.h
+++ b/plasma/generic/dataengines/powermanagement/powermanagementengine.h
@@ -63,7 +63,7 @@ private:
 
     QStringList m_sources;
 
-    QHash<QString, QString> m_batterySources;
+    QHash<QString, QString> m_batterySources;  // <udi, Battery0>
 
 };
 
